internal/agentcfg/elasticsearch.go

```diff
@@ -0,0 +1,267 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package agentcfg
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/monitoring"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+const ElasticsearchIndexName = ".apm-agent-configuration"
+
+const (
+	// ErrInfrastructureNotReady is returned when a fetch request comes in while
+	// the first run of refresh cache has not completed yet.
+	ErrInfrastructureNotReady = "agentcfg infrastructure is not ready"
+
+	// ErrNoValidElasticsearchConfig is an error where the server is
+	// not properly configured to fetch agent configuration.
+	ErrNoValidElasticsearchConfig = "no valid elasticsearch config to fetch agent config"
+)
+
+const refreshCacheTimeout = 5 * time.Second
+
+type ElasticsearchFetcher struct {
+	client          *elasticsearch.Client
+	cacheDuration   time.Duration
+	fallbackFetcher Fetcher
+
+	mu    sync.RWMutex
+	last  time.Time
+	cache []AgentConfig
+
+	searchSize int
+
+	firstRunCompleted atomic.Bool
+	invalidESCfg      atomic.Bool
+	cacheInitialized  atomic.Bool
+
+	logger *logp.Logger
+
+	metrics fetcherMetrics
+}
+
+type fetcherMetrics struct {
+	fetchES, fetchFallback, fetchFallbackUnavailable,
+	cacheRefreshSuccesses, cacheRefreshFailures,
+	cacheEntriesCount atomic.Int64
+}
+
+func NewElasticsearchFetcher(client *elasticsearch.Client, cacheDuration time.Duration, fetcher Fetcher) *ElasticsearchFetcher {
+	logger := logp.NewLogger("agentcfg")
+	return &ElasticsearchFetcher{
+		client:          client,
+		cacheDuration:   cacheDuration,
+		fallbackFetcher: fetcher,
+		searchSize:      100,
+		logger:          logger,
+	}
+}
+
+// Fetch finds a matching agent config based on the received query.
+func (f *ElasticsearchFetcher) Fetch(ctx context.Context, query Query) (Result, error) {
+	if f.cacheInitialized.Load() {
+		// Happy path: serve fetch requests using an initialized cache.
+		f.mu.RLock()
+		defer f.mu.RUnlock()
+		defer f.metrics.fetchES.Add(1)
+		return matchAgentConfig(query, f.cache), nil
+	}
+
+	if !f.firstRunCompleted.Load() {
+		f.logger.Warnf("rejecting fetch request: first refresh cache run has not completed")
+		return Result{}, errors.New(ErrInfrastructureNotReady)
+	}
+
+	if f.fallbackFetcher != nil {
+		defer f.metrics.fetchFallback.Add(1)
+		return f.fallbackFetcher.Fetch(ctx, query)
+	}
+
+	if f.invalidESCfg.Load() {
+		defer f.metrics.fetchFallbackUnavailable.Add(1)
+		f.logger.Errorf("rejecting fetch request: no valid elasticsearch config")
+		return Result{}, errors.New(ErrNoValidElasticsearchConfig)
+	} else {
```

Comment:
I did the same to the rest to remove the nested `if`s in the decision tree but left this because this was not nested. But I agree it will be better to remove the else.